File: src/downloader/parsers.py
# downloader/parsers.py
"""Functions to extract DOIs from various academic citation file formats."""

import json
import re
from collections.abc import Callable
from pathlib import Path
from typing import Any

import bibtexparser
import defusedxml.ElementTree as ET
import rispy

from .utils import clean_doi
from .pdf_extractor import extract_dois_from_pdf

DOI_REGEX = r"\b(10[.]\d{4,9}/[-._;()/:A-Z0-9]+)\b"


def _parse_generic(text: str, loader: Callable[[str], Any], key: str = "doi") -> list[str]:
    """Generic parser for structured formats."""
    dois = set()
    try:
        data = loader(text)
        # Handle both list of dicts (RIS, JSON) and object with entries (BibTeX)
        items = data.entries if hasattr(data, "entries") else data
        for item in items:
            val = item.get(key)
            if val and (cleaned := clean_doi(val)):
                dois.add(cleaned)
    except Exception:
        pass
    return list(dois)


def _parse_bibtex(text: str) -> list[str]:
    """Extracts DOIs from a BibTeX string."""
    return _parse_generic(text, bibtexparser.loads, "doi")


def _parse_ris(text: str) -> list[str]:
    """Extracts DOIs from an RIS string using the rispy library."""
    return _parse_generic(text, rispy.loads, "doi")


def _parse_endnote_xml(text: str) -> list[str]:
    """Extracts DOIs from an EndNote XML string."""
    dois = set()
    try:
        root = ET.fromstring(text)
        for doi_node in root.findall(".//electronic-resource-num"):
            if doi_node.text and (cleaned := clean_doi(doi_node.text)):
                dois.add(cleaned)
    except Exception:
        pass
    return list(dois)


def _parse_json(text: str) -> list[str]:
    """Extracts DOIs from a Zotero-style JSON export."""
    return _parse_generic(text, json.loads, "DOI")


def _parse_plain_text(text: str) -> list[str]:
    """Finds all DOIs in a plain text or CSV string using regex."""
    dois = set()
    for match in re.findall(DOI_REGEX, text, re.IGNORECASE):
        if cleaned := clean_doi(match):
            dois.add(cleaned)
    return list(dois)


def _detect_parser_from_content(text: str) -> Callable[[str], list[str]]:
    """Detects the appropriate parser based on file content."""
    if "TY  -" in text and "ER  -" in text:
        return _parse_ris
    elif "@article" in text.lower() or "@book" in text.lower():
        return _parse_bibtex
    return _parse_plain_text


def extract_dois_from_file(filepath: str) -> list[str]:
    """
    Reads a file and extracts DOIs based on its content and extension.
    Supports .bib, .ris, .xml, .enw, .txt, .csv, .json, and .pdf.
    """
    p = Path(filepath)
    if not p.exists():
        raise FileNotFoundError(f"File not found: {filepath}")

    ext = p.suffix.lower()

    if ext == ".pdf":
        return extract_dois_from_pdf(filepath)

    text = p.read_text(encoding="utf-8", errors="ignore")

    parser_map = {
        ".bib": _parse_bibtex,
        ".ris": _parse_ris,
        ".xml": _parse_endnote_xml,
        ".enw": _parse_endnote_xml,
        ".json": _parse_json,
    }

    if ext in parser_map:
        dois = parser_map[ext](text)
    elif ext in [".txt", ".csv"]:
        parser = _detect_parser_from_content(text)
        dois = parser(text)
    else:
        dois = _parse_plain_text(text)

    return sorted(list(set(dois)))

File: internal/parsers/parsers.go
package parsers

import (
	"encoding/json"
	"encoding/xml"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"litnexus/internal/utils"
)

var (
	bibtexDOIRegex = regexp.MustCompile(`(?i)doi\s*=\s*[{"]([^}"]+)[}"]`)
	risDOIRegex    = regexp.MustCompile(`(?i)^DO\s*-\s*(.+)$`)
)

// ExtractDOIsFromFile reads a file and extracts all valid DOIs.
func ExtractDOIsFromFile(path string) ([]string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read file %s: %w", path, err)
	}

	ext := strings.ToLower(filepath.Ext(path))
	text := string(data)

	var dois []string
	switch ext {
	case ".bib":
		dois = parseBibTeX(text)
	case ".ris":
		dois = parseRIS(text)
	case ".xml", ".enw":
		dois = parseEndNoteXML(data)
	case ".json":
		dois = parseJSON(data)
	case ".txt", ".csv":
		dois = parsePlainTextOrAuto(text)
	default:
		dois = utils.ExtractDOIsFromText(text)
	}

	return deduplicateAndSort(dois), nil
}

func parseBibTeX(text string) []string {
	matches := bibtexDOIRegex.FindAllStringSubmatch(text, -1)
	var result []string
	for _, m := range matches {
		if len(m) > 1 {
			if cleaned := utils.CleanDOI(m[1]); cleaned != "" {
				result = append(result, cleaned)
			}
		}
	}

	// Fallback to regex text extraction if bibtex field match was sparse
	if len(result) == 0 {
		result = utils.ExtractDOIsFromText(text)
	}
	return result
}

func parseRIS(text string) []string {
	lines := strings.Split(text, "\n")
	var result []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if m := risDOIRegex.FindStringSubmatch(line); len(m) > 1 {
			if cleaned := utils.CleanDOI(m[1]); cleaned != "" {
				result = append(result, cleaned)
			}
		}
	}

	if len(result) == 0 {
		result = utils.ExtractDOIsFromText(text)
	}
	return result
}

type endNoteXMLRecord struct {
	ElectronicResourceNum string `xml:"electronic-resource-num"`
}

type endNoteXMLRoot struct {
	Records []endNoteXMLRecord `xml:"records>record"`
}

func parseEndNoteXML(data []byte) []string {
	var root endNoteXMLRoot
	var result []string
	if err := xml.Unmarshal(data, &root); err == nil {
		for _, rec := range root.Records {
			if cleaned := utils.CleanDOI(rec.ElectronicResourceNum); cleaned != "" {
				result = append(result, cleaned)
			}
		}
	}

	if len(result) == 0 {
		result = utils.ExtractDOIsFromText(string(data))
	}
	return result
}

func parseJSON(data []byte) []string {
	var items []map[string]interface{}
	var result []string

	if err := json.Unmarshal(data, &items); err == nil {
		for _, item := range items {
			for k, v := range item {
				if strings.EqualFold(k, "doi") {
					if strVal, ok := v.(string); ok {
						if cleaned := utils.CleanDOI(strVal); cleaned != "" {
							result = append(result, cleaned)
						}
					}
				}
			}
		}
	}

	if len(result) == 0 {
		result = utils.ExtractDOIsFromText(string(data))
	}
	return result
}

func parsePlainTextOrAuto(text string) []string {
	if strings.Contains(text, "TY  -") && strings.Contains(text, "ER  -") {
		return parseRIS(text)
	}
	if strings.Contains(strings.ToLower(text), "@article") || strings.Contains(strings.ToLower(text), "@book") {
		return parseBibTeX(text)
	}
	return utils.ExtractDOIsFromText(text)
}

func deduplicateAndSort(dois []string) []string {
	seen := make(map[string]bool)
	var unique []string
	for _, d := range dois {
		if !seen[d] {
			seen[d] = true
			unique = append(unique, d)
		}
	}
	sort.Strings(unique)
	return unique
}
